File: BackEnd/src/main/java/com/arsansys/RemaPartners/services/ProductoService.java
package com.arsansys.RemaPartners.services;

import java.util.List;

import org.springframework.stereotype.Service;

import com.arsansys.RemaPartners.models.entities.ProductoEntity;
import com.arsansys.RemaPartners.models.enums.EEstado;
import com.arsansys.RemaPartners.models.enums.EMoneda;

@Service
public interface ProductoService {

    /**
     * Obtiene la lista de todos los productos.
     * 
     * @return Lista de entidades de producto.
     */
    abstract List<ProductoEntity> getProductos();

    /**
     * Obtiene un producto por su ID.
     * 
     * @param id ID del producto.
     * @return Entidad de producto encontrada o null si no existe.
     */
    abstract ProductoEntity getProductoById(String id);

    /**
     * Obtiene productos por su estado.
     * 
     * @param estado Estado del producto.
     * @return Lista de productos con el estado especificado.
     */
    abstract List<ProductoEntity> getProductosByEstado(EEstado estado);

    /**
     * Obtiene productos por su moneda.
     * 
     * @param moneda Moneda del producto.
     * @return Lista de productos con la moneda especificada.
     */
    abstract List<ProductoEntity> getProductosByMoneda(EMoneda moneda);

    /**
     * Obtiene productos por el ID de la categoría.
     * 
     * @param idCategoria ID de la categoría.
     * @return Lista de productos de la categoría.
     */
    abstract List<ProductoEntity> getProductosByIdCategoria(String idCategoria);

    /**
     * Obtiene productos por la marca.
     * 
     * @param marca Marca del producto.
     * @return Lista de productos de la marca.
     */
    abstract List<ProductoEntity> getProductosByMarca(String marca);

    /**
     * Obtiene productos por el modelo.
     * 
     * @param modelo Modelo del producto.
     * @return Lista de productos del modelo.
     */
    abstract List<ProductoEntity> getProductosByModelo(String modelo);

    /**
     * Obtiene productos por el título.
     * 
     * @param titulo Título del producto.
     * @return Lista de productos con el título especificado.
     */
    abstract List<ProductoEntity> getProductosByTitulo(String titulo);

    /**
     * Obtiene productos por la descripción.
     * 
     * @param descripcion Descripción del producto.
     * @return Lista de productos con la descripción especificada.
     */
    abstract List<ProductoEntity> getProductosByDescripcion(String descripcion);

    /**
     * Crea un nuevo producto.
     * 
     * @param productoEntity Entidad de producto a crear.
     * @return Entidad de producto creada.
     */
    abstract ProductoEntity createProducto(ProductoEntity productoEntity);

    /**
     * Actualiza un producto existente.
     * 
     * @param productoEntity Entidad de producto con los datos actualizados.
     * @return Entidad de producto actualizada.
     */
    abstract ProductoEntity updateProducto(ProductoEntity productoEntity);

    /**
     * Elimina un producto por su ID.
     * 
     * @param id ID del producto a eliminar.
     */
    abstract void deleteProductoById(String id);

    /**
     * Cambia el estado (activo/inactivo) de un producto.
     * 
     * @param id ID del producto.
     */
    abstract void toggleStatus(String id);

}

File: FrontEnd/rema-partners/src/services/producto.service.ts
import axios from "axios";
import Cookies from "js-cookie";
import type { Producto, ProductoModify } from "@/models/producto";

const API_BASE_URL = "http://localhost:8080";
const CATEGORIA_API_URL = `${API_BASE_URL}/vendedor/producto`;

export const productoService = {
    /**
     * Cambia el estado (activo/inactivo) de un producto.
     * @param {string} id - ID del producto.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async toggleStatus(id: string) {
        return axios.post(`${CATEGORIA_API_URL}/toggleStatus/${id}`);
    },
    /**
     * Crea un nuevo producto.
     * @param {ProductoModify} producto - Datos del producto a crear.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async createProducto(producto: ProductoModify) {
        return axios.post(`${CATEGORIA_API_URL}/create`, producto);
    },
    /**
     * Obtiene todos los productos.
     * @returns {Promise} Promesa con la lista de productos.
     */
    async getProductos() {
        return axios.get(`${CATEGORIA_API_URL}/getAll`);
    },
    /**
     * Obtiene un producto por su ID.
     * @param {string} id - ID del producto.
     * @returns {Promise} Promesa con los datos del producto.
     */
    async getProductoById(id: string) {
        return axios.get(`${CATEGORIA_API_URL}/getById/${id}`);
    },
    /**
     * Actualiza un producto existente.
     * @param {string} id - ID del producto.
     * @param {ProductoModify} producto - Datos actualizados del producto.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async updateProducto(id: string, producto: ProductoModify) {
        return axios.post(`${CATEGORIA_API_URL}/update/${id}`, producto);
    },
    /**
     * Elimina un producto por su ID.
     * @param {string} id - ID del producto.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async deleteProducto(id: string) {
        return axios.delete(`${CATEGORIA_API_URL}/delete/${id}`);
    },
    /**
     * Obtiene los productos de un usuario por su ID.
     * @param {string} id - ID del usuario.
     * @returns {Promise} Promesa con la lista de productos.
     */
    async getProductosByUsuario(id: string) {
        return axios.get(`${CATEGORIA_API_URL}/getByUserId/${id}`);
    },
    /**
     * Obtiene los productos por ID de categoría.
     * @param {string} id - ID de la categoría.
     * @returns {Promise} Promesa con la lista de productos.
     */
    async getProductosByIdCategoria(id: string) {
        return axios.get(`${CATEGORIA_API_URL}/getByIdCategoria/${id}`);
    },
    /**
     * Marca un producto como vendido y actualiza el stock.
     * @param {string} id - ID del producto.
     * @param {number} quantity - Cantidad vendida.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async markAsSold(id: string, quantity: number) {
        const response = await axios.get(`${CATEGORIA_API_URL}/getById/${id}`);
        const productData = response.data;

        // Update the product stock by reducing the quantity sold
        productData.stock = Math.max(0, productData.stock - quantity);
        return axios.post(`${CATEGORIA_API_URL}/update/${id}`, productData);
    }
}